include/umf/mempolicy.h

```diff
@@ -0,0 +1,38 @@
+/*
+ *
+ * Copyright (C) 2023 Intel Corporation
+ *
+ * Under the Apache License v2.0 with LLVM Exceptions. See LICENSE.TXT.
+ * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+ *
+ */
+
+#ifndef UMF_MEMPOLICY_H
+#define UMF_MEMPOLICY_H 1
+
+#include <umf/base.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef struct umf_mempolicy_t *umf_mempolicy_handle_t;
+
+#define UMF_MEMPOLICY_INTERLEAVE _UMF_MEMPOLICY_INTERLEAVE()
+#define UMF_MEMPOLICY_BIND _UMF_MEMPOLICY_BIND()
+#define UMF_MEMPOLICY_PREFERED _UMF_MEMPOLICY_PREFERED()
+
+const umf_mempolicy_handle_t _UMF_MEMPOLICY_INTERLEAVE(void);
+const umf_mempolicy_handle_t _UMF_MEMPOLICY_BIND(void);
+const umf_mempolicy_handle_t _UMF_MEMPOLICY_PREFERED(void);
+
+umf_result_t umfMempolicyCreateInterleave(size_t part_size,
+                                          umf_mempolicy_handle_t *policy);
+
+umf_result_t umfMempolicyDestroy(umf_mempolicy_handle_t policy);
```

Comment:
I think a discussion is needed before introducing the new API. Do we have a set of use cases listed somewhere to review if the proposed API conforms to it?

The questions I have:

1. Why not use an approach similar to what we have with memory providers config structure? E.g.:
```
enum umf_bind_type_t {
    UMF_BIND_INTERLEAVE,
    UMF_BIND_BIND,
    UMF_BIND_PREFERED,
    ...
};
umf_result_t umfMempolicyCreate(umf_bind_type_t bind_type, umf_mempolicy_handle_t *policy);
umf_result_t umfMempolicyDestroy(umf_mempolicy_handle_t policy);
umf_result_t imfMempolicySetPartSize(size_t part_size);
...
```

2. Do we need Create/Destroy API at all? Can just a data structure describe the policy?
```
struct umf_mempolicy_t {
    umf_bind_type_t bind_type;
    size_t part_size;
    ...
};
```

Reply:
@vinser52  We do have an issue for memspace and policies discussion here: https://github.com/oneapi-src/unified-memory-framework/issues/230 but it does not list the use cases.

Reply:
This implementation is based on design and examples from your repo @vinser52 - i was not a part of the team when it was created, so I'm not sure how mature it is. 
Regarding to your questions:
0. First off all this is aligned with other API - so if we want to change it we have to change it everywhere in UMF. 
1. I was thinking about this - this approach is more user friendly - for basic use cases you don't have to create an extra object. You have to create one only for "advanced cases". 
2. Having opaque structs give us flexibility regarding further extensions, without braking an backward compatibility. If we put an struct in public header, we cannot change it's size, or remove fields. Of course there is workaround:
```c 
struct umf_mempolicy_t {
    umf_bind_type_t bind_type;
    size_t part_size;
    char x[300] // reserved for futher use.
}
```
But i prefer opaque struct in this case, as it is more flexible.

Reply:
> This implementation is based on design and examples from your repo @vinser52 - i was not a part of the team when it was created, so I'm not sure how mature it is.

It is immature.

> Having opaque structs give us flexibility regarding further extensions, without braking a backward compatibility. If we put an struct in public header, we cannot change it's size, nor remove fields. 

Ok, I agree that an opaque structure is preferable. 

> I was thinking about this - this approach is more user friendly - for basic use cases you don't have to create an extra object. You have to create one only for "advanced cases".

What do you mean under basic cases? I think this approach is aligned with pool/provider configs.

Regarding the API proposed in this PR, my concern is the extensibility. Consider in the future we need to support more parameters in addition to `part_size`. Approach with setter APIs (`umf_result_t umfMempolicySetPartSize(size_t part_size);`) for each parameter is more flexible I think.

I don't particularly appreciate having API functions for particular binding type. E.g. :
```
umfMempolicyCreateInterleave(size_t part_size, umf_mempolicy_handle_t *policy)
```

I prefer having a single pair of create/destroy functions for mempolicy.


With @bratpiorka we agreed to discuss this on our Monday sync. Please join if you can.

Reply:
Done